### PR TITLE
bugfix: unnecessary `/` appended to s3 path joins

### LINKED ISSUE #2575: S3 ListObjects broken with Aws\EndpointV2
### Describe the bug

When using an `endpoint` + `bucket_endpoint` with `Aws\EndpointV2` the `listObjects` method of an S3 Client doesn't seem to work as expected. It results in a double-slash URL:

```
GET https://bucket-name.endpoint.dev/bucket-name//?prefix=%2F&encoding-type=url
```

Removing the trailing slash from the `endpoint` argument resolves the error, however the returned results are always empty, as the bucket name path is completely stripped from the effective URL:

```
https://bucket-name.endpoint.dev/?prefix=%2F&encoding-type=url
```

### Expected Behavior

The `ListObjects` call should work with a trailing slash on the endpoint and return the items from the S3 bucket.

### Current Behavior

`ListObjects` produces an invalid URL.

### Reproduction Steps

```php
use Aws\Endpoint\PartitionEndpointProvider;
use Aws\S3\S3Client;

require 'vendor/autoload.php';

$s3Client = new S3Client( [
	'version' => '2006-03-01',
	'endpoint' => 'https://bucket-name.endpoint.dev/bucket-name/',
	'signature' => 'v4',
	'bucket_endpoint' => true,
	'region' => 'us-east-1',
	'credentials' => [
		'key' => 'admin',
		'secret' => 'password',
	],
] );

var_dump( $s3Client->listObjects( [
	'Bucket' => 'bucket-name',
	'Prefix' => '/',
] ) );
```

### Possible Solution

I do not have suggestions on how to fix `ListObjects` with `Aws\EndpointV2`, but current workarounds are:

* Use `ListObjectsV2` instead
* Use `PartitionEndpointProvider::defaultProvider()` for `endpoint_provider`

### Additional Information/Context

_No response_

### SDK version used

3.246.1

### Environment details (Version of PHP (`php -v`)? OS name and version, etc.)

PHP 8.0.24

### PR DESCRIPTION
**Pull request by stobrien89**
*Issue #, if available:*
closes #2575, revision to #2572

*Description of changes:*
Update to S3 model pre-processing to remove unnecessary trailing slashes when joining paths.  

Updates to #2573 to disallow overrides of any endpoint parameter built-in values (this should be handled by contextparams) except in the case of  the legacy `'@use_dual_stack_endpoint', '@use_accelerate_endpoint'` and `'@use_path_style_endpoint' ` command options

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/EndpointV2/EndpointV2SerializerTrait.php [hunk 1]
```diff
@@ -127,28 +127,25 @@ private function filterEndpointCommandArgs(
...
-                    && strpos($builtIn, 'SDK') === 0
-                ) {
+                if (!empty($value->getBuiltIn())) {
                     continue;
                 }
                 $filteredArgs[$name] = $commandArgs[$name];
             }
         }
+
+        if ($this->api->getServiceName() === 's3') {
```
Comment (SamRemis) [inline src/EndpointV2/EndpointV2SerializerTrait.php:141]: Is there no way to take this out of the shared files?

File: src/S3/S3Client.php [hunk 2]
```diff
@@ -678,9 +678,11 @@ private function processEndpointV2Model()
         foreach($definition['operations'] as &$operation) {
             if (isset($operation['http']['requestUri'])) {
                 $requestUri = $operation['http']['requestUri'];
-                $requestUri = str_replace('/{Bucket}', '/', $requestUri);
-                $requestUri = str_replace('//', '/', $requestUri);
-                $requestUri = str_replace('/?', '?', $requestUri);
+                if ($requestUri === "/{Bucket}") {
+                    $requestUri = str_replace('/{Bucket}', '/', $requestUri);
+                } else {
+                    $requestUri = str_replace('/{Bucket}', '', $requestUri);
```
Comment (SamRemis) [inline src/S3/S3Client.php:684]: just trying to undrstand this- why do we only have a trailing slash when the uri is /{Bucket}?

File: src/Api/Serializer/RestSerializer.php [hunk 3]
```diff
@@ -231,20 +231,24 @@ function (array $matches) use ($varDefinitions) {
...
         $path = $this->endpoint->getPath();
-        //accounts for removal of bucket from requestUri path
+
+        //Accounts for trailing '/' in path when custom endpoint
+        //is provided to endpointProviderV2
         if ($this->api->isModifiedModel()
             && $this->api->getServiceName() === 's3'
         ) {
-            $relative = $path . $relative;
+            substr($path, -1) === '/' && substr($relative, 0, 1) === '/'
```
Comment (SamRemis) [inline src/Api/Serializer/RestSerializer.php:241]: this just removes it from the relative if the path has a trailing slash and the relative has a trailing?  I feel like there's different logic that could accomplish the same thing; maybe an if statement would be clearer here?  Like in pseudo-code:
```
if (substr($path, -1) === '/' && substr($relative, 0, 1) === '/') {
$path = rtrim($path, '/');
}
$relative = $path . $relative;
```
Of course this won't work if the $path can't be changed if it's used later, but it looks like it's not.  

Also, maybe change `substr($relative, 0, 1)` to `$relative[0]`

File: src/S3/S3Client.php (hunk 2 above)
Comment (stobrien89) [reply src/S3/S3Client.php:684]: We/S3 expect there to be a trailing slash on operations (CreateBucket, DeleteBucket, etc.) where there is no additional path or query.  For the others, if there's more path string beyond `{Bucket}`, a forward slash will already be present, or the query will just be appended.

**Review by SamRemis — APPROVED** [review]

**Pull request merged**